### PR TITLE
[GSK-2892] Improve RAG toolset

### PR DESCRIPTION
**Pull request by pierlj**

### CODE REVIEW

File: giskard/rag/question_generators/base.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            logger.debug("JSON decoding error, trying to fix the JSON string.")
+
+        logger.debug("Raw output: %s", raw_json)
+        # Let's see if it's just a matter of markdown format (```json ... ```)
+        match = re.search(r"```json\s{0,5}(.*?)\s{0,5}```", raw_json, re.DOTALL)
+        if match:
+            try:
+                return json.loads(match.group(1), strict=False)
+            except json.JSONDecodeError:
+                pass
```
Comment (Hartorn) [inline giskard/rag/question_generators/base.py:64]: Would be nice to get another debugger log here

**Review by Hartorn — COMMENTED** [review]
Overall solid PR, but too much repeated used of string key in dict. 
This could be improved by using dataclass/constant/formatting function

File: giskard/rag/knowledge_base.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+                n_neighbors=50,
+                min_dist=0.5,
+                n_components=2,
+            )
+            self._reduced_embeddings_inst = reducer.fit_transform(self._embeddings)
+        return self._reduced_embeddings_inst
+
+    @property
+    def _dimension(self):
+        return self._embeddings[0].shape[0]
```
[the remark this answers is not in the document]
Comment (pierlj) [reply giskard/rag/knowledge_base.py:198]: It's not be possible to have empty embeddings

File: giskard/rag/knowledge_base.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
+from typing import Optional, Sequence
+
+import logging
+
+import numpy as np
+import pandas as pd
+from sklearn.cluster import HDBSCAN
+
+from ..datasets.metadata.text_metadata_provider import _detect_lang
```
[the remark this answers is not in the document]
Comment (pierlj) [reply giskard/rag/knowledge_base.py:9]: I think so, it's only used in `text_metadata_provider.py`

**Pull request merged**